src/libPersistence/ContractStorage2.cpp

```diff
@@ -342,6 +364,58 @@ bool ContractStorage2::FetchStateValue(const dev::h160& addr, const bytes& src,
   return SerializeToArray(value, dst, 0);
 }
 
+bool ContractStorage2::FetchExternalStateValue(
+    const dev::h160& caller, const dev::h160& target, const bytes& src,
+    unsigned int s_offset, bytes& dst, unsigned int d_offset, bool& foundVal,
+    string& type, uint32_t caller_version) {
+  // get caller version if not available
+  if (caller_version == std::numeric_limits<uint32_t>::max()) {
+    std::map<std::string, bytes> t_caller_version;
+    string version_key =
+        GenerateStorageKey(caller, SCILLA_VERSION_INDICATOR, {});
```

Comment:
I'm surprised we need scilla_version here. It looks like it's used to store several version of the contract state - is that to make it easier to deal with future migrations?

Reply:
I believe we based it on the remote state specs where, yes, the version is also included: https://github.com/Zilliqa/scilla-compiler/wiki/Interaction-with-Blockchain#fetching-states-of-other-contracts

Reply:
"...a fail happens when
...
- The version of the external contract is not the same as the contract being executed."

That's not going to work. If someone deploys a contract that other contracts are supposed to interact with, then we can't expect the owner of that contract to update the contract (i.e., deploy a new one and migrate the contract state) when we update to a new Scilla version, nor can we expect the interacting contracts to stay at the earlier Scilla version (which we may choose to no longer support).

Reply:
@jjcnn I recall writing this point because we wanted to disallow interaction b/w contracts of different versions. Shouldn't we have this anymore?

Reply:
We have no restriction whatsoever on the versions of the sender and the recipient of messages, so I doubt we made that decision - we either made the opposite decision, or didn't decide on anything yet. 

In any case I really don't think it will work to restrict contract interaction like that. The result would be that contracts will exist in isolated silos based on their version number. If a token contract is written in version 0, then users of that token would forever be restricted to using Scilla version 0 to trade in those tokens, and that isn't a viable solution in the long-term.

Reply:
In any case, I suggest we ignore this point as far as this PR is concerned and if we finalize on allowing cross-version remote state reads, I'll implement that as a separate PR. 

The reason I prefer to do that is: this PR is a re-do of earlier PR with some parts removed. It's confusing as it is to keep track. So I'd prefer it be as near to the original one as we practically can, and do further fixes as a separate PR.

Reply:
That's fine - it doesn't make any difference as long as there's only one Scilla version in play.